relayer/pkg/chainlink/keys/keystore.go

```
@@ -22,4 +58,117 @@ type NonceManager interface {
+	}
+	return k, nil
+}
+
+// Sign implements the LoopKeystore interface
+// this implementation wraps starknet specific curve and expects
+// hash: byte representation (big-endian) of *big.Int.
+func (ks *MemKeystore) Sign(id string, hash []byte) ([]byte, error) {
+	k, err := ks.Get(id)
+	if err != nil {
+		return nil, err
+	}
+
+	starkHash := new(big.Int).SetBytes(hash)
+
+	x, y, err := caigo.Curve.Sign(starkHash, k)
+	if err != nil {
+		return nil, fmt.Errorf("error signing data with curve: %w", err)
+	}
+
+	s := &signature{
+		x: x,
+		y: y,
+	}
+	return s.bytes()
+}
+
+// signature is an intermediate representation for translating between a raw-bytes signature and a caigo
+// signature comprised of big.Ints
+type signature struct {
+	x, y *big.Int
+}
+
+func (s *signature) bytes() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	n, err := buf.Write(padBytes(s.x.Bytes(), maxPointByteLen))
+	if err != nil {
+		return nil, fmt.Errorf("error writing 'x' component of signature: %w", err)
+	}
+	if n != maxPointByteLen {
+		return nil, fmt.Errorf("unexpected write length of 'x' component of signature: wrote %d expected %d", n, maxPointByteLen)
+	}
+
+	n, err = buf.Write(padBytes(s.y.Bytes(), maxPointByteLen))
+	if err != nil {
+		return nil, fmt.Errorf("error writing 'y' component of signature: %w", err)
+	}
+	if n != maxPointByteLen {
+		return nil, fmt.Errorf("unexpected write length of 'y' component of signature: wrote %d expected %d", n, maxPointByteLen)
+	}
+
+	if buf.Len() != signatureLen {
+		return nil, fmt.Errorf("error in signature length")
+	}
+	return buf.Bytes(), nil
+}
+
+// b is expected to encode x,y components in accordance with [signature.bytes]
+func signatureFromBytes(b []byte) (*signature, error) {
+	if len(b) != signatureLen {
+		return nil, fmt.Errorf("expected slice len %d got %d", signatureLen, len(b))
+	}
+	x := b[:maxPointByteLen]
+	y := b[maxPointByteLen:]
+
+	return &signature{
+		x: new(big.Int).SetBytes(x),
+		y: new(big.Int).SetBytes(y),
```

Can these ints be negative? b/c `Bytes()` doesn't preserve the sign. I ran in to this with LOOPs and ended up using a protobuf https://github.com/smartcontractkit/chainlink-relay/blob/main/pkg/loop/internal/pb/big.go

---

that's a good question. 

yes, based on the elliptic curve, they can be negative.

the current OCR key signing for starknet is using big.Int.Bytes, and i largely copied that.

@aalu1418 what is the expectation regarding the sign of the points on the curve -- is the OCR implementation consciously choosing to ignore/drop it, or is that an oversight in the implementation? 

https://github.com/smartcontractkit/chainlink-starknet/blob/develop/relayer/pkg/chainlink/keys/ocr2key.go

---

for the OCR key - i remember that being a conscious decision to only use the positive values

however, for the transmitter key/signatures I'm not sure - since we have previously passed the private key caigo handled the signs / encodings internally 🤔

---

i looked at the official python implementation. in my reading, all the int values are positive. can one of you read the python and/or caigo give your assessment? cc @archseer 

i'm also going to ask research on the general topic of big.Int.Set/Bytes and elliptic curves.

the private key is positve     https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L103

the msgHash is positive https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L139


the signature components --
r is positive:         https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L156

s is positive: 

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L170

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/signature.py#L111

https://github.com/starkware-libs/starkex-for-spot-trading/blob/607f0b4ce507e1d95cd018d206a2797f6ba4aab4/src/starkware/crypto/starkware/crypto/signature/math_utils.py#L52

---

Would it make sense to sanity check and panic if negative?

---

yes, good idea. i think the right place for that is the `Decode` implementation, which i've now added

---

Where is the sanity check? I see `(s *signature) bytes()` checking length, but not sign.

---

bump ^

---

Oh right, i meant to reply to this.

It turns out that this is not needed. it's not possible to check that they are negative -- the SetBytes assumes an absolute value, so you would need a different en/decoding to represent a negative number and therefore you can't even check for negatives in a meaningful way here.

the same idea applies to the bytes() func: since it's using big.Int.Bytes(), it is not possible to encode negative values.

the only place a negative could be represented is in line 110-112. but there again the value is coming from the caigo.Sign, which returns positive values. 

we could add a panic there i suppose. wdyt?

i've written tests for the round trip in keystore and signature. do you see anything missing in them?

---

If `(s *signature) bytes()` is called with a signature that has negative values, it will silently drop that information. IMHO this is where the panic should occur: at the narrowest scope without making assumptions about anything else.